conda_forge_tick/git_utils.py

```diff
@@ -88,77 +97,533 @@ def github_client() -> github.Github:
+    def __str__(self):
+        return str(self.value)
 
-    Parameters
-    ----------
-    feedstock_name : str
-        The feedstock without '-feedstock'.
+    INFINITY = math.inf
+    """
+    Python does not have support for a literal infinity type, so we use this enum for it.
+    """
 
-    Returns
-    -------
-    branch : str
-        The default branch (e.g., 'main').
+
+class GitConnectionMode(enum.StrEnum):
+    SSH = "ssh"
+    HTTPS = "https"
+
+
+class GitCliError(Exception):
+    pass
+
+
+class RepositoryNotFoundError(Exception):
+    """
+    Raised when a repository is not found.
     """
-    return (
-        github_client()
-        .get_repo(f"conda-forge/{feedstock_name}-feedstock")
-        .default_branch
-    )
 
+    pass
 
-def get_github_api_requests_left() -> Union[int, None]:
-    """Get the number of remaining GitHub API requests.
 
-    Returns
-    -------
-    left : int or None
-        The number of remaining requests, or None if there is an exception.
+class GitCli:
     """
-    gh = github3_client()
-    try:
-        left = gh.rate_limit()["resources"]["core"]["remaining"]
-    except Exception:
-        left = None
+    A simple wrapper around the git command line interface.
+    """
+
+    @staticmethod
+    def _run_git_command(
+        cmd: list[str | Path],
+        working_directory: Path | None = None,
+        check_error: bool = True,
+    ) -> subprocess.CompletedProcess:
+        """
+        Run a git command.
+        :param cmd: The command to run, as a list of strings.
+        :param working_directory: The directory to run the command in. If None, the command will be run in the current
+        working directory.
+        :param check_error: If True, raise a GitCliError if the git command fails.
+        :return: The result of the git command.
+        """
+        git_command = ["git"] + cmd
+
+        logger.debug(f"Running git command: {git_command}")
 
-    return left
+        try:
+            return subprocess.run(git_command, check=check_error, cwd=working_directory)
+        except subprocess.CalledProcessError as e:
+            raise GitCliError("Error running git command.") from e
+
+    def reset_hard(self, git_dir: Path, to_treeish: str = "HEAD"):
+        """
+        Reset the git index of a directory to the state of the last commit with `git reset --hard HEAD`.
+        :param git_dir: The directory to reset.
+        :param to_treeish: The treeish to reset to. Defaults to "HEAD".
+        :raises GitCliError: If the git command fails.
+        """
+        self._run_git_command(["reset", "--quiet", "--hard", to_treeish], git_dir)
+
+    def clone_repo(self, origin_url: str, target_dir: Path):
+        """
+        Clone a Git repository. If target_dir already exists, it will be reset to the state of the git index
+        with `git reset --hard HEAD`. This may be unexpected behavior because target_dir could contain a
+        different repository. Use with caution.
```

Comment:
If you did this at the CLI, git would raise an error. We should do the same I think.

Reply:
Agreed.

Reply:
The logic is now moved into clone_fork_and_branch